### ISSUE TITLE
Delete Capitol Corridor data that was incorrectly tagged as AC Transit

### ISSUE THREAD

**Issue by lauriemerrell** [cal-itp/data-infra#1018]
See #1017. We have been downloading Capitol Corridor (`calitp_itp_id` 56) data and assigning it to both Capitol Corridor *and* to AC Transit (`calitp_itp_id` 4). 

We should:
- [x] Confirm the dates for which the Capitol Corridor data was downloaded under the AC Transit ID
- [ ] Confirm that the data was in fact downloaded in both locations (i.e., there are no dates for which data was downloaded only to AC Transit and not to Capitol Corridor)
- [ ] Assuming the above is confirmed, delete the erroneous AC Transit copies of the data -- **there is both schedule and RT data that was miscategorized**
- [x] Wait until #1024 merges because until that is fixed, the incorrect data will continue populating.

**Comment by lauriemerrell** [comment]
Doing research/prep for this:

- Appears that the issue was introduced in 2aea1114f494eac3e6156c4deca5203e4ae9f255, so all the data in the following locations would be incorrect -- in each case would run a `gsutil mv <junkyard>/<old path>` to move the data out of the production data and somewhere else -- we should decide on what `<junkyard>` is:
   - `gs://gtfs-data/schedule/2021-12-*/4_1/*` and `gs://gtfs-data/schedule/2022-*/4_1/*`
       -  doing two wildcard matches is faster than globbing everything AND see issue #521 -- there was Sage Stage data assigned to this ID/URL combo earlier in 2021. I think we need to handle that data separately because I'm not sure it was duplicated anywhere else (assume it needs to be reassigned to Sage Stage, and not copied away to the junkyard)
   - `gs://gtfs-data/schedule/processed/2021-12-*_4_1/*` and `gs://gtfs-data/schedule/processed/2022-*_4_1/*`
   - `gs://gtfs-data/rt/*/4/1/*`
       -  Definitely will need to break this one up -- it is very slow
   - There does not seem to be `rt-processed` data for `4` or for `56` -- I guess their `vehicle_positions` data is empty? 
- It seems like the erroneous data didn't actually make it into `views` because(?) there was already erroneous data as described in #521... `SELECT * FROM cal-itp-data-infra.views.gtfs_schedule_dim_feeds WHERE calitp_itp_id = 4 AND calitp_url_number = 1` yields only Sage Stage data, and nothing from Capitol Corridor, but `SELECT * FROM cal-itp-data-infra.gtfs_schedule_history.calitp_files_updates WHERE calitp_itp_id = 4 AND calitp_url_number = 1` does find both Sage Stage and Capitol Corridor data. So we may want to make a separate ticket to investigate why prior erroneous data blocked the loading of new data -- in particular, I am wondering if #1024 won't be sufficient to actually get the new, correct AC Transit data all the way to `views` because of the old Sage Stage data that's hanging around.

**Comment by hunterowens** [comment]
Perhaps ahead of this worth implementing backups and a recovery procedure
for this bucket?

**Comment by evansiroky** [comment]
### Response to @lauriemerrell

> * Appears that the issue was introduced in [2aea111](https://github.com/cal-itp/data-infra/commit/2aea1114f494eac3e6156c4deca5203e4ae9f255), 

dang it. Although I seem to remember that I force-pushed some old branch so it might not be entirely my fault!

> so all the data in the following locations would be incorrect -- in each case would run a `gsutil mv <old path> <junkyard>/<old path>` to move the data out of the production data and somewhere else -- we should decide on what `<junkyard>` is:
>   
>   * `gs://gtfs-data/schedule/2021-12-*/4_1/*` and `gs://gtfs-data/schedule/2022-*/4_1/*`
>     
>     * doing two wildcard matches is faster than globbing everything AND see issue [Bug: Sage Stage routes appear in gtfs_schedule.routes when filtering for AC Transit #521](https://github.com/cal-itp/data-infra/issues/521) -- there was Sage Stage data assigned to this ID/URL combo earlier in 2021. I think we need to handle that data separately because I'm not sure it was duplicated anywhere else (assume it needs to be reassigned to Sage Stage, and not copied away to the junkyard)
>   * `gs://gtfs-data/schedule/processed/2021-12-*_4_1/*` and `gs://gtfs-data/schedule/processed/2022-*_4_1/*`
>   * `gs://gtfs-data/rt/*/4/1/*`
>     
>     * Definitely will need to break this one up -- it is very slow
>   * There does not seem to be `rt-processed` data for `4` or for `56` -- I guess their `vehicle_positions` data is empty?

We may also want to investigate whether the contents of `gs://gtfs-data/schedule/*-*-*/feeds_raw.csv` and `gs://gtfs-data/schedule/*-*-*/feeds.csv` should be changed.

> * It seems like the erroneous data didn't actually make it into `views` because(?) there was already erroneous data as described in [Bug: Sage Stage routes appear in gtfs_schedule.routes when filtering for AC Transit #521](https://github.com/cal-itp/data-infra/issues/521)... `SELECT * FROM cal-itp-data-infra.views.gtfs_schedule_dim_feeds WHERE calitp_itp_id = 4 AND calitp_url_number = 1` yields only Sage Stage data, and nothing from Capitol Corridor, but `SELECT * FROM cal-itp-data-infra.gtfs_schedule_history.calitp_files_updates WHERE calitp_itp_id = 4 AND calitp_url_number = 1` does find both Sage Stage and Capitol Corridor data. So we may want to make a separate ticket to investigate why prior erroneous data blocked the loading of new data -- in particular, I am wondering if [Use correct AC transit 511 agency key #1024](https://github.com/cal-itp/data-infra/pull/1024) won't be sufficient to actually get the new, correct AC Transit data all the way to `views` because of the old Sage Stage data that's hanging around.

In terms of the theory of how this data ingestion pipeline works, each of the BigQuery datasets that don't directly rely on raw data (in this pipeline, that would be the agencies.yml - or more specifically the resulting `feeds.csv` and `feeds_raw.csv` files and downloaded static feed files) should be able to be recreated in entirety from the raw data. Therefore, we should be looking into tracing all these discrepancies as far back as possible in the pipeline and making changes accordingly.

### Response to @hunterowens 

> Perhaps ahead of this worth implementing backups and a recovery procedure for this bucket?

I think that'd be a good idea.

### One other idea

One other idea I had is another option to go about this is to have some kind of "corrections" data and DAG task logic that we can use to run additional processing to filter out historical mistakes in the raw data before further ingestion into the pipeline. Like in this case, we could have some kind of data that says that ITP ID 4 / url_number 1 is invalid between these dates and then there'd be a DAG task(s) that would hide/remove that data from being exposed to the rest of the pipeline.

**Comment by lauriemerrell** [comment]
Thank you @hunterowens and @evansiroky! Backup makes sense, and yes `feeds.csv` will need to be updated too. 

#### Re: DAG task to suppress data
I think it's an interesting idea, but I am a bit nervous about maintaining something like that for cases like this where it was our fault (not the data producer's fault) that the data was "wrong" (i.e., I think a DAG to suppress might make more sense to address data corruption issues in the underlying feed where the feed is correctly categorized; in cases like this and #521 where we have misfiled the data, I don't think there's value in leaving the raw data in the incorrect location). I do think that writing some kind of "delete data" script that's standardized would make sense as opposed to literally running ad hoc `gsutil mv` commands. This could also allow us to handle the `feeds.csv` considerations.

#### Re: why the Capitol Corridor data didn't make it into `views`
I am officially confused and will need to investigate more. It looks like the fixed AC Transit data **did** make it in last night (`SELECT * FROM cal-itp-data-infra.views.gtfs_schedule_dim_feeds WHERE calitp_itp_id = 4 AND calitp_url_number = 1` now yields one Sage Stage entry and one AC Transit entry 👍). So something else must have prevented the Capitol Corridor data from populating. 

`SELECT * FROM cal-itp-data-infra.gtfs_schedule_type2.calitp_feeds WHERE calitp_itp_id = 4 AND calitp_url_number = 1` does **not** turn up the Capitol Corridor data, but doing that same query in `cal-itp-data-infra.gtfs_schedule_type2.agency_clean` **does**. So, something odd happened here. I will look into it further.

**Comment by themightychris** [comment]
> Perhaps ahead of this worth implementing backups and a recovery procedure for this bucket?

GCP's Object Versioning is currently off for this bucket, turning it on might be an option but we'll need to consider the storage implications of enabling that for this whole bucket. Maybe it would make more sense in conjunction with splitting the content up over multiple buckets a bit. We should survey the data in that bucket and its size/turnover

**Comment by lauriemerrell** [comment]
TLDR: There is an issue in [`gtfs_views_staging.calitp_feeds`](https://github.com/cal-itp/data-infra/blob/main/airflow/dags/gtfs_views_staging/calitp_feeds.sql) that means that in any case where we have an ID/URL combination in `agencies.yaml`, that combination is removed for some period of time, and then it's added back again, the re-added data will not appear in `views` until it's updated a second time. 

Actual notes on this specific situation:

The Capitol Corridor data is present in all the relevant `gtfs_schedule_history` data. It falls out of the pipeline at the creation of [`gtfs_views_staging.calitp_feeds`](https://github.com/cal-itp/data-infra/blob/main/airflow/dags/gtfs_views_staging/calitp_feeds.sql). 

Checking each of the subqueries:
- `gtfs_schedule_feed_snapshot` - CC data is present
- `feed_id_cross_date` - CC data is present
- `lag_md5_hash` - CC data is present -- correctly shows data deleted in May when Sage Stage issue was first fixed, data missing May - Dec, then data resumes in Dec
- `hash_check` - CC data is present but the `is_changed` flag is `null` on `2021-12-29` (the date that the CC data starts being collected under this ID) and false subsequently, until it is `true` on `2022-02-02` (the date that the assignment was fixed)
- `feed_snapshot_with_removed` is where the CC data falls out -- the `is_removed` flag stays `true` up until `2022-12-28`, correctly, but then there is nothing until `2022-02-02`.

So, the problem is in [`hash_check`](https://github.com/cal-itp/data-infra/blob/main/airflow/dags/gtfs_views_staging/calitp_feeds.sql#L51-L56). Basically, the logic here didn't account for the case where an ID/URL combo is present, then is removed for some period of time, and then is reinstated. 

Specifically, `previous_hash` is `null` on `2021-12-29` and `is_first_extraction` is also `false`, so `is_changed` is not populated for this row. Weirdly `calitp_hash!=prev_calitp_hash` does not trigger as `true` when one is null and the other is populated? 

I think we could add a hacky check for `OR ((calitp_hash IS NULL) <> (prev_calitp_hash IS NULL))` or something here... 

The question is whether we think we actually need to support this case: **Is there a legitimate situation where data would be removed and then reinstated?** If we remove the underlying raw data from the wrong location, I think this would clear up on its own. **But** if we have cases like this again where something was done in error in the past, that error was fixed, and then new correct data is added, the correct data will not be picked up immediately. (Like, in this case it's just a fluke that the re-added data was also wrong.)

### `gtfs_schedule_history` notes
In all cases I'm checking for rows with `calitp_itp_id` = 4 and `calitp_url_number` = 1
- `gtfs_schedule_history.calitp_feeds` has 36 rows. All but one are the erroneous data (the 511 agency ID is `AM`). The remaining row is the corrected data presumably from yesterday. There is no Sage Stage data here (I think this table was created in August and maybe not backpopulated?)
- `cal-itp-data-infra.gtfs_schedule_history.calitp_feed_updates` has 4 entries, 1 in April (Sage Stage), 1 in December (CC), 1 in January (CC) and then yesterday's correct AC Transit extract. 
-   `gtfs_schedule_history.calitp_status` has 64 rows, seemingly encompassing every single day that this combination was defined in `agencies.yaml` (the Sage Stage era, the CC era, and now the corrected era.)

**Comment by atvaccaro** [comment]
This is no longer necessary as the backfill was performed with URLs present in status.csv files.

**Comment by atvaccaro** [comment]
This is RT not schedule

**Comment by lauriemerrell** [comment]
Closing as will not be fixed because we decided not to backfill RT data prior to September 2022 (per Data Services Stakeholder Sync meeting on 6/6/23) and the misclassified RT data is not available anywhere.
